modules/arkit/arkit_interface.mm

```diff
@@ -0,0 +1,711 @@
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "camera_ios.h"
+#include "core/os/input.h"
+#include "core/os/os.h"
+#include "scene/resources/surface_tool.h"
+#include "servers/camera/camera_feed.h"
+#include "servers/visual/visual_server_globals.h"
+
+#import <ARKit/ARKit.h>
+#import <UIKit/UIKit.h>
+
+#include "arkit_interface.h"
+#include "arkit_session_delegate.h"
+
+// just a dirty workaround for now, declare these as globals. I'll probably encapsulate ARSession and associated logic into an mm object and change ARKitInterface to a normal cpp object that consumes it.
+ARSession *ar_session;
+ARKitSessionDelegate *ar_delegate;
+NSTimeInterval last_timestamp;
+
+/* this is called when we initialize or when we come back from having our app pushed to the background, just (re)start our session */
+void ARKitInterface::start_session() {
+	// We're active...
+	session_was_started = true;
+
+	// Ignore this if we're not initialized...
+	if (initialized) {
+		print_line("Starting ar session");
+		ARWorldTrackingConfiguration *configuration = [ARWorldTrackingConfiguration new];
+		configuration.lightEstimationEnabled = light_estimation_is_enabled;
+		configuration.planeDetection = plane_detection_is_enabled ? ARPlaneDetectionVertical | ARPlaneDetectionHorizontal : 0;
+
+		// make sure our camera is on
+		CameraFeed *feed = CameraIOS::get_arkit_feed();
+		feed->set_active(true);
+
+		[ar_session runWithConfiguration:configuration];
+	}
+}
+
+void ARKitInterface::stop_session() {
+	session_was_started = false;
+
+	// Ignore this if we're not initialized...
+	if (initialized) {
+		// make sure our camera is off
+		CameraFeed *feed = CameraIOS::get_arkit_feed();
+		feed->set_active(false);
+
+		[ar_session pause];
+	}
+}
+
+void ARKitInterface::_notification(int p_what) {
+	// TODO, this is not being called, need to find out why, possibly because this is not a node.
+	// in that case we need to find a way to get these notifications!
+	switch (p_what) {
+		case MainLoop::NOTIFICATION_WM_FOCUS_IN: {
+			print_line("Focus in");
+
+			start_session();
+		}; break;
+		case MainLoop::NOTIFICATION_WM_FOCUS_OUT: {
+			print_line("Focus out");
+
+			stop_session();
+		}; break;
+		default:
+			break;
+	}
+}
+
+bool ARKitInterface::get_anchor_detection_is_enabled() const {
+	return plane_detection_is_enabled;
+}
+
+void ARKitInterface::set_anchor_detection_is_enabled(bool p_enable) {
+	if (plane_detection_is_enabled != p_enable) {
+		plane_detection_is_enabled = p_enable;
+
+		// Restart our session (this will be ignore if we're not initialised)
```

Comment:
The comment says `initalised` but the code reads `session_was_started`. Which is correct?

Reply:
Let me come back to you on that one once I have a moment to catch up :)

Reply:
Ok, so this is some weirdness around how phones work, it doesn't work fully yet because I haven't got notifications to work properly just yet.

Anyway, the problem lies around an app becoming dormant when you switch away from it. Basically ARKit needs to be doing its thing when you've initialized it AND we have focus.

So if ARKit is initialized before we obtain focus, we initialize ARKit but we don't start the session. Then when we obtain focus start_session is called and we start our ARKit session.
If we obtain focus before ARKit is initialized start_session is called but initialize is false so we don't do anything more, the session_was_started ensures that once we initialize we also setup the ARKit session.

If at any time after this has all been setup we loose focus stop_session is called and we simply pause our ARKit session and unpause it in start_session once focus is regained (the unpauze is actually a reinitialize as settings on the phone may have changed).

Reply:
I wonder if we should either hook the `appDidResignActive` and `appDidBecomeActive` events from iOS, or if we should hoist them up to Godot and provide other notifications for the window.

Reply:
They are now hooked up, these result in FOCUS IN and FOCUS OUT notifications and after I managed to get notifications to work with the ARVRServer the other day and that change was merged in, should now work with ARKit :)